tidb-binlog/driver/reader/kafka_consumer/kafka_go.go

```diff
@@ -0,0 +1,227 @@
+			MaxBytes:  10e6, // 1MB
+		}),
+		conn:   conn,
+		ctx:    ctx,
+		cancel: cancel,
+	}, nil
+}
+
+// ConsumerFromOffset implements Consumer.ConsumerFromOffset
+func (k *KafkaGO) ConsumeFromOffset(offset int64, consumerChan chan<- *KafkaMsg, done <-chan struct{}) error {
+	err := k.client.SetOffset(offset)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		select {
+		case <-done:
+			log.Info("finish consumer")
+			return nil
+		default:
+			ctx, cancel := context.WithTimeout(k.ctx, KafkaWaitTimeout)
+			kmsg, err := k.client.ReadMessage(ctx)
+			cancel()
+			if err != nil {
+				log.Warn("kafka-go consume from offset failed",
+					zap.Int64("offset", k.client.Offset()),
+					zap.Error(err))
+				continue
+			}
+			msg := &KafkaMsg{
+				Value:  kmsg.Value,
+				Offset: kmsg.Offset,
+			}
+			consumerChan <- msg
+		}
+	}
+}
+
+// SeekOffsetFromTS implements Consumer.SeekOffsetFromTS
+func (k *KafkaGO) SeekOffsetFromTS(ts int64, topic string, partitions []int32) ([]int64, error) {
+	if len(partitions) == 0 {
+		pts, err := k.conn.ReadPartitions(topic)
+		if err != nil {
+			log.Error("get partitions from topic failed", zap.String("topic", topic), zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+		for _, pt := range pts {
+			partitions = append(partitions, int32(pt.ID))
+		}
+	}
+	return k.seekOffsets(topic, partitions, ts)
+}
+
+// seekOffsets returns all valid offsets in partitions
+func (k *KafkaGO) seekOffsets(topic string, partitions []int32, pos int64) ([]int64, error) {
+	offsets := make([]int64, len(partitions))
+	for _, partition := range partitions {
+		start, err := k.conn.ReadFirstOffset()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		end, err := k.conn.ReadLastOffset()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("seek offsets in",
+			zap.String("topic", topic),
+			zap.Int32("partition", partition),
+			zap.Int64("start", start),
+			zap.Int64("end", end),
+			zap.Int64("target ts", pos))
+
+		offset, err := k.seekOffset(topic, partition, start, end-1, pos)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		log.Info("seek offset success", zap.Int64("offset", offset), zap.Int64("target ts", pos))
+		offsets[partition] = offset
+	}
+
+	return offsets, nil
+}
+
+func (k *KafkaGO) seekOffset(topic string, partition int32, start int64, end int64, ts int64) (offset int64, err error) {
+	startTS, err := k.getTSAtOffset(topic, partition, start)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	if ts < startTS {
+		log.Warn("given ts is smaller than oldest message's ts, some binlogs may lose", zap.Int64("given ts", ts), zap.Int64("oldest ts", startTS))
+		offset = start
+		return
+	} else if ts == startTS {
+		offset = start + 1
+		return
+	}
+
+	for start < end {
+		mid := (end-start)/2 + start
+		var midTS int64
+		midTS, err = k.getTSAtOffset(topic, partition, mid)
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		if midTS <= ts {
```

Comment:
if `midTS == ts`, can it return directly?

Reply:
Yes,[ it can ](https://github.com/pingcap/tidb-tools/pull/292/files#diff-8bd56c363bb9394e8480c72ccafb2211R198), same as [sarama](https://github.com/pingcap/tidb-tools/pull/292/files#diff-8de333e66d7d8d94c619760775d51273R179)